### PR TITLE
CSHARP-3061: Clarify how a driver must handle wrong set name in single topology.

### PR DESCRIPTION
**Pull request by DmitryLukyanov**
EG: https://evergreen.mongodb.com/version/5ea76fd25623435bc78830bc

### CODE REVIEW

File: tests/MongoDB.Driver.Tests/MongoClientSettingsTests.cs [hunk 1]
```diff
@@ -493,7 +493,6 @@ public void TestFromUrlWithMongoDBX509_without_username()
         }
 
         [Theory]
-        [InlineData("mongodb+srv://username:password@test5.test.build.10gen.cc/?connect=direct", ConnectionStringScheme.MongoDB, "localhost.test.build.10gen.cc:27017")]
```
Comment (DmitryLukyanov) [inline tests/MongoDB.Driver.Tests/MongoClientSettingsTests.cs:496]: This test fails with the new requirement regarding directConnection. They're described here:https://github.com/mongodb/specifications/commit/e56f5eceed7729f8b9b43a4a1f76c7e5840db49f#diff-5e688c46cf9e7c28e7b4b821cf4cd145R66

File: src/MongoDB.Driver.Core/Core/Clusters/SingleServerCluster.cs [hunk 2]
```diff
@@ -177,6 +179,16 @@ private void ServerDescriptionChanged(object sender, ServerDescriptionChangedEve
             var newServerDescription = args.NewServerDescription;
             var newClusterDescription = Description;
 
+            if (_replicaSetName != null && newServerDescription.ReplicaSetConfig != null)
```
Comment (rstam) [inline src/MongoDB.Driver.Core/Core/Clusters/SingleServerCluster.cs]: What if `_replicaSetName` is not null but `newServerDescription.ReplicaSetConfig` IS null?

Shouldn't we set the `ServerType` to `Unknown` in that case also?

Comment (rstam) [reply src/MongoDB.Driver.Core/Core/Clusters/SingleServerCluster.cs]: I propose we refactor this new code as:

    if (_replicaSetName != null)
    {
        var replicaSetConfig = newServerDescription.ReplicaSetConfig;
        if (replicaSetConfig == null || replicaSetConfig.Name != _replicaSetName)
        {
            // if the replica set name does not match then the ServerType in the ServerDescription MUST be replaced with Unknown
            newServerDescription = newServerDescription.With(type: ServerType.Unknown);
        }
    }

Note: the comment has been updated also

File: src/MongoDB.Driver.Core/Core/Configuration/ConnectionString.cs [hunk 3]
```diff
@@ -807,6 +808,12 @@ private void Parse()
             ExtractScheme(match);
             ExtractHosts(match);
 
+            if (_connect.HasValue && _directConnection.HasValue)
+            {
+                throw new MongoConfigurationException("Specifying both connect and directConnection is invalid.");
```
Comment (rstam) [inline src/MongoDB.Driver.Core/Core/Configuration/ConnectionString.cs]: Suggested error message change:

    throw new MongoConfigurationException("Connect and directConnection cannot both be specified.");

File: src/MongoDB.Driver.Core/Core/Configuration/ConnectionString.cs [hunk 4]
```diff
@@ -818,6 +825,16 @@ private void Parse()
                     "Specifying both tlsInsecure and tlsDisableCertificateRevocationCheck is invalid.");
             }
 
+            if (_scheme == ConnectionStringScheme.MongoDBPlusSrv && _connect == ClusterConnectionMode.Direct)
+            {
+                throw new MongoConfigurationException("Direct connect is not allowed to be used with SRV.");
```
Comment (rstam) [inline src/MongoDB.Driver.Core/Core/Configuration/ConnectionString.cs]: Suggested error message change:

     throw new MongoConfigurationException("Direct connect cannot be used with SRV.");

File: src/MongoDB.Driver.Core/Core/Configuration/ConnectionString.cs [hunk 5]
```diff
@@ -818,6 +825,16 @@ private void Parse()
...
+            }
+
+            if (_hosts.Count > 1 && _connect == ClusterConnectionMode.Direct)
+            {
+                throw new MongoConfigurationException("Direct connect is not allowed to be used with multiple seeds.");
```
Comment (rstam) [inline src/MongoDB.Driver.Core/Core/Configuration/ConnectionString.cs]: Suggested error message change:

    throw new MongoConfigurationException("Direct connect cannot be used with multiple host names.");

File: src/MongoDB.Driver.Core/Core/Configuration/ConnectionString.cs [hunk 6]
```diff
@@ -1199,6 +1219,25 @@ private bool EnsureTlsInsecureIsValid(bool value)
             return value;
         }
 
+        private ClusterConnectionMode? GetEffectiveConnectionMode(ClusterConnectionMode? mode, bool? directConnection, string replicaSet)
```
Comment (rstam) [inline src/MongoDB.Driver.Core/Core/Configuration/ConnectionString.cs]: Change `mode` parameter name to `connect` (to match the _connect field it corresponds to).

File: tests/MongoDB.Driver.Core.Tests/Core/Clusters/SingleServerClusterTests.cs [hunk 7]
```diff
@@ -148,18 +149,44 @@ public void Dispose_should_dispose_of_the_server()
             _capturedEvents.Any().Should().BeFalse();
         }
 
+        [Theory]
+        [ParameterAttributeData]
+        public void ServerDescription_type_should_be_replaced_on_Unknown_when_isMaster_setName_wrong(
```
Comment (rstam) [inline tests/MongoDB.Driver.Core.Tests/Core/Clusters/SingleServerClusterTests.cs]: Rename method to:

    ServerDescription_type_should_be_replaced_with_Unknown_when_isMaster_setName_is_different

File: tests/MongoDB.Driver.Core.Tests/Core/Configuration/ConnectionStringTests.cs [hunk 8]
```diff
@@ -514,6 +514,59 @@ public void When_a_database_name_is_specified(string connectionString, string db
...
+        [InlineData("mongodb://localhost/?directConnection=true&replicaSet=yeah", true, ClusterConnectionMode.Direct)]
+        [InlineData("mongodb://localhost/?directConnection=true", true, ClusterConnectionMode.Direct)]
+        [InlineData("mongodb://localhost/?directConnection=false&replicaSet=yeah", false, ClusterConnectionMode.ReplicaSet)]
+        [InlineData("mongodb://localhost/?directConnection=false", false, ClusterConnectionMode.Automatic)]
+        public void When_a_directConenction_is_specified(string connectionString, bool directConnection, ClusterConnectionMode mode)
+        {
+            var subject = new ConnectionString(connectionString);
+
+            var result = bool.Parse(subject.GetOption("directConnection"));
+            result.Should().Be(directConnection);
```
Comment (rstam) [inline tests/MongoDB.Driver.Core.Tests/Core/Configuration/ConnectionStringTests.cs]: Change assertions to:

    subject.Connect.Should().Be(connect);
    subject._directConnection().Should().Be(directConnection);

Note: requires adding a `ConnectionStringReflector` class (unless you add a `DirectConnection` public property).

    public static class ConnectionStringReflector
    {
        public static bool _directConnection(this ConnectionString obj) => (bool)Reflector.GetFieldValue(obj, nameof(_directConnection));
    }

File: tests/MongoDB.Driver.Core.Tests/Core/Configuration/ConnectionStringTests.cs [hunk 9]
```diff
@@ -514,6 +514,59 @@ public void When_a_database_name_is_specified(string connectionString, string db
...
+        }
+
+        [Theory]
+        [InlineData("mongodb+srv://localhost/?directConnection=false", false)]
+        [InlineData("mongodb+srv://localhost/?directConnection=true", true)]
+        public void When_a_directConnection_is_specified_with_a_srv_scheme(string connectionString, bool shouldThrow)
+        {
+            ConnectionString result = null;
```
Comment (rstam) [inline tests/MongoDB.Driver.Core.Tests/Core/Configuration/ConnectionStringTests.cs]: Rename `result` to `subject`.

File: tests/MongoDB.Driver.Core.Tests/Core/Configuration/ConnectionStringTests.cs [hunk 10]
```diff
@@ -514,6 +514,59 @@ public void When_a_database_name_is_specified(string connectionString, string db
...
+
+            if (shouldThrow)
+            {
+                exception.Should().BeOfType<MongoConfigurationException>();
+            }
+            else
+            {
+                exception.Should().BeNull();
+                var directConnection = bool.Parse(result.GetOption("directConnection"));
+                directConnection.Should().BeFalse();
```
Comment (rstam) [inline tests/MongoDB.Driver.Core.Tests/Core/Configuration/ConnectionStringTests.cs]: Change assertions to:

    exception.Should().BeNull();
    subject._directConnection().Should().BeFalse();

File: tests/MongoDB.Driver.Core.Tests/Core/Configuration/ConnectionStringTests.cs [hunk 11]
```diff
@@ -514,6 +514,59 @@ public void When_a_database_name_is_specified(string connectionString, string db
...
+            }
+        }
+
+        [Theory]
+        [InlineData("mongodb://localhost1,localhost2/?directConnection=false", false)]
+        [InlineData("mongodb://localhost1,localhost2/?directConnection=true", true)]
+        public void When_a_directConnection_is_specified_with_multiple_seeds(string connectionString, bool shouldThrow)
+        {
+            ConnectionString result = null;
```
Comment (rstam) [inline tests/MongoDB.Driver.Core.Tests/Core/Configuration/ConnectionStringTests.cs]: Rename `result` to `subject`.

File: tests/MongoDB.Driver.Core.Tests/Core/Configuration/ConnectionStringTests.cs [hunk 12]
```diff
@@ -514,6 +514,59 @@ public void When_a_database_name_is_specified(string connectionString, string db
...
+
+            if (shouldThrow)
+            {
+                exception.Should().BeOfType<MongoConfigurationException>();
+            }
+            else
+            {
+                exception.Should().BeNull();
+                var directConnection = bool.Parse(result.GetOption("directConnection"));
+                directConnection.Should().BeFalse();
```
Comment (rstam) [inline tests/MongoDB.Driver.Core.Tests/Core/Configuration/ConnectionStringTests.cs]: Replace assertions with:

    exception.Should().BeNull();
    subject._directConnection().Should().BeFalse();

File: tests/MongoDB.Driver.Core.Tests/Core/Configuration/ConnectionStringTests.cs [hunk 13]
```diff
@@ -514,6 +514,59 @@ public void When_a_database_name_is_specified(string connectionString, string db
             subject.DatabaseName.Should().Be(db);
         }
 
+        [Theory]
+        [InlineData("mongodb://localhost/?directConnection=true&replicaSet=yeah", true, ClusterConnectionMode.Direct)]
+        [InlineData("mongodb://localhost/?directConnection=true", true, ClusterConnectionMode.Direct)]
+        [InlineData("mongodb://localhost/?directConnection=false&replicaSet=yeah", false, ClusterConnectionMode.ReplicaSet)]
+        [InlineData("mongodb://localhost/?directConnection=false", false, ClusterConnectionMode.Automatic)]
+        public void When_a_directConenction_is_specified(string connectionString, bool directConnection, ClusterConnectionMode mode)
```
Comment (rstam) [inline tests/MongoDB.Driver.Core.Tests/Core/Configuration/ConnectionStringTests.cs]: Rename `mode` parameter to `connect`.

File: tests/MongoDB.Driver.Core.Tests/Core/Configuration/ConnectionStringTests.cs [hunk 14]
```diff
@@ -514,6 +514,59 @@ public void When_a_database_name_is_specified(string connectionString, string db
...
+                exception.Should().BeNull();
+                var directConnection = bool.Parse(result.GetOption("directConnection"));
+                directConnection.Should().BeFalse();
+            }
+        }
+
+        [Theory]
+        [InlineData("mongodb://localhost1,localhost2/?directConnection=false", false)]
+        [InlineData("mongodb://localhost1,localhost2/?directConnection=true", true)]
+        public void When_a_directConnection_is_specified_with_multiple_seeds(string connectionString, bool shouldThrow)
```
Comment (rstam) [inline tests/MongoDB.Driver.Core.Tests/Core/Configuration/ConnectionStringTests.cs]: Rename method to:

    When_a_directConnection_is_specified_with_multiple_hosts

"seeds" only applies to replica sets. Multiple hosts is more generic (e.g. it applies to shard routers also).

**Review by rstam — CHANGES_REQUESTED** [review]
In general looks good. Some minor changes requested.

File: src/MongoDB.Driver.Core/Core/Configuration/ConnectionString.cs (hunk 3 above)
Comment (DmitryLukyanov) [reply src/MongoDB.Driver.Core/Core/Configuration/ConnectionString.cs]: done

File: src/MongoDB.Driver.Core/Core/Configuration/ConnectionString.cs (hunk 4 above)
Comment (DmitryLukyanov) [reply src/MongoDB.Driver.Core/Core/Configuration/ConnectionString.cs]: done

File: src/MongoDB.Driver.Core/Core/Configuration/ConnectionString.cs (hunk 5 above)
Comment (DmitryLukyanov) [reply src/MongoDB.Driver.Core/Core/Configuration/ConnectionString.cs]: done

File: src/MongoDB.Driver.Core/Core/Configuration/ConnectionString.cs (hunk 6 above)
Comment (DmitryLukyanov) [reply src/MongoDB.Driver.Core/Core/Configuration/ConnectionString.cs]: done

File: tests/MongoDB.Driver.Core.Tests/Core/Clusters/SingleServerClusterTests.cs (hunk 7 above)
Comment (DmitryLukyanov) [reply tests/MongoDB.Driver.Core.Tests/Core/Clusters/SingleServerClusterTests.cs]: done

File: tests/MongoDB.Driver.Core.Tests/Core/Configuration/ConnectionStringTests.cs (hunk 8 above)
Comment (DmitryLukyanov) [reply tests/MongoDB.Driver.Core.Tests/Core/Configuration/ConnectionStringTests.cs]: done

File: tests/MongoDB.Driver.Core.Tests/Core/Configuration/ConnectionStringTests.cs (hunk 9 above)
Comment (DmitryLukyanov) [reply tests/MongoDB.Driver.Core.Tests/Core/Configuration/ConnectionStringTests.cs]: done

File: tests/MongoDB.Driver.Core.Tests/Core/Configuration/ConnectionStringTests.cs (hunk 14 above)
Comment (DmitryLukyanov) [reply tests/MongoDB.Driver.Core.Tests/Core/Configuration/ConnectionStringTests.cs]: done

File: tests/MongoDB.Driver.Core.Tests/Core/Configuration/ConnectionStringTests.cs (hunk 11 above)
Comment (DmitryLukyanov) [reply tests/MongoDB.Driver.Core.Tests/Core/Configuration/ConnectionStringTests.cs]: done

File: tests/MongoDB.Driver.Core.Tests/Core/Configuration/ConnectionStringTests.cs (hunk 12 above)
Comment (DmitryLukyanov) [reply tests/MongoDB.Driver.Core.Tests/Core/Configuration/ConnectionStringTests.cs]: done

File: tests/MongoDB.Driver.Core.Tests/Core/Configuration/ConnectionStringTests.cs (hunk 13 above)
Comment (DmitryLukyanov) [reply tests/MongoDB.Driver.Core.Tests/Core/Configuration/ConnectionStringTests.cs]: done

File: src/MongoDB.Driver.Core/Core/Clusters/SingleServerCluster.cs (hunk 2 above)
Comment (DmitryLukyanov) [reply src/MongoDB.Driver.Core/Core/Clusters/SingleServerCluster.cs]: done

Comment (DmitryLukyanov) [reply src/MongoDB.Driver.Core/Core/Clusters/SingleServerCluster.cs]: it took time to understand that ReplicaSetConfig is presented only on the driver-side

**Comment by DmitryLukyanov** [comment]
EG: https://evergreen.mongodb.com/version/5ea85d4d9ccd4e57ae2259eb
ready for review

**Review by rstam — APPROVED** [review]
LGTM

**Review by jyemin — APPROVED** [review]
LGTM

**Pull request closed without merging**